my/core/__main__.py

```diff
@@ -360,69 +329,121 @@ def module_install(args: Namespace) -> None:
-        opsi = ops.add_parser('install', help='Install module dependencies')
-        add_module_arg(opsi)
-        opsi.add_argument('--user', action='store_true', help='same as pip --user')
-        opsi.set_defaults(func=module_install)
-        # todo could add functions to check specific module etc..
+    MODULE is one or more specific module names to check (e.g. my.reddit)
+    Otherwise, checks all modules
+    '''
+    if not skip_conf:
+        config_ok()
+    # TODO check that it finds private modules too?
+    modules_check(verbose, list_all, quick, list(module))
 
-    return p
 
+@main.group(name='config', short_help='work with configuration')
+def config_grp() -> None:
+    '''Act on your HPI configuration'''
+    pass
 
-def main() -> None:
-    p = parser()
-    args = p.parse_args()
 
-    func = getattr(args, 'func', None)
-    if func is None:
-        p.print_help()
-        sys.exit(1)
+@config_grp.command(name='check', short_help='check config')
+def config_check_cmd() -> None:
+    '''Check your HPI configuration file'''
+    ok = config_ok()
+    sys.exit(0 if ok else False)
 
-    import tempfile
-    with tempfile.TemporaryDirectory() as td:
-        # cd into tmp dir to prevent accidental imports..
-        os.chdir(str(td))
-        func(args)
+
+@config_grp.command(name='create', short_help='create user config')
+def config_create_cmd() -> None:
+    '''Create user configuration file for HPI'''
+    config_create()
+
+
+@main.command(name='modules', short_help='list available modules')
+@click.option('--all', 'list_all', is_flag=True, help='List all modules, including disabled')
+def module_cmd(list_all: bool) -> None:
+    '''List available modules'''
+    list_modules(list_all)
+
+
+@main.group(name='module', short_help='module management')
+def module_grp() -> None:
+    '''Module management'''
+    pass
+
+
+@module_grp.command(name='requires', short_help='print module reqs')
+@click.argument('MODULE')
+def module_requires_cmd(module: str) -> None:
+    '''
+    Print MODULE requirements
+
+    MODULE is a specific module name (e.g. my.reddit)
+    '''
+    module_requires(module)
+
+
+@module_grp.command(name='install', short_help='install module deps')
+@click.option('--user', is_flag=True, help='same as pip --user')
+@click.argument('MODULE')
+def module_install_cmd(user: bool, module: str) -> None:
+    '''
+    Install dependencies for a module using pip
+
+    MODULE is a specific module name (e.g. my.reddit)
+    '''
+    # todo could add functions to check specific module etc..
+    module_install(user, module)
+
+
+# todo: add more tests?
+# its standard click practice to have the function click calls be a separate
+# function from the decorated function, as it allows the application-specific code to be
+# more testable. also allows hpi commands to be imported and called manually from
+# other python code
+
+
+def test_requires() -> None:
+    from click.testing import CliRunner
```

Comment:
> added a test using click.testing.CliRunner (tests the CLI in an isolated environment)

Neat! Although I've started wondering how it could possibly work since something like cgroups isn't portable and generally would need root permissions etc -- seems that it's pretty basic, just hacks the env variables & IO streams
https://github.com/pallets/click/blob/c4542263a96dd067ed1ded12a83553da05bb61dd/src/click/testing.py#L172

And the filesystem isolation is merely a chdir https://github.com/pallets/click/blob/c4542263a96dd067ed1ded12a83553da05bb61dd/src/click/testing.py#L384 Also isn't running by default, which is weird, arguably one of the most important reasons to isolate? 

I might contribute to their docs, feel like it can be misleading if people don't look in the source.
But otherwise still super convenient if one is aware of the limitations.